### PR TITLE
Expose AppDomain.PermissionSet

### PR DESCRIPTION
**Pull request by maryamariyan**
Fixes: https://github.com/dotnet/corefx/issues/32745

cc: @danmosemsft @ericstj @ViktorHofer @safern

### CODE REVIEW

File: src/System.Security.Permissions/src/System.Security.Permissions.csproj [hunk 1]
```diff
@@ -172,6 +169,11 @@
     <Compile Include="System\Transactions\DistributedTransactionPermission.cs" />
     <Compile Include="System\Transactions\DistributedTransactionPermissionAttribute.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetGroup)' == 'netstandard'">
```
Comment (jkotas) [inline src/System.Security.Permissions/src/System.Security.Permissions.csproj]: Do we really need to be building netstandard version of this?

File: src/System.Security.Permissions/ref/System.Security.Permissions.netcoreapp.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (danmosemsft) [inline src/System.Security.Permissions/ref/System.Security.Permissions.netcoreapp.cs]: blank line

File: src/System.Security.Permissions/src/System.Security.Permissions.csproj (hunk 1 above)
Comment (danmosemsft) [reply src/System.Security.Permissions/src/System.Security.Permissions.csproj]: @ericstj ?

Comment (safern) [reply src/System.Security.Permissions/src/System.Security.Permissions.csproj]: Yes we need it for other frameworks like UAP. All of our packages need to have a matching ref and src implementation for NS in order to be restorable (used).

Comment (jkotas) [reply src/System.Security.Permissions/src/System.Security.Permissions.csproj]: This change should be done for UAP as well (it is partially done in this PR already). We structure UAP and NETCOREAPP same way. UAP should not be using netstandard version of the package after this change.

File: src/System.Security.Permissions/ref/System.Security.Permissions.csproj [hunk 3]
```diff
@@ -6,6 +6,8 @@
   </PropertyGroup>
   <ItemGroup>
     <Compile Include="System.Security.Permissions.cs" />
+    <Compile Include="System.Security.Permissions.$(TargetGroup).cs" />
```
Comment (danmosemsft) [inline src/System.Security.Permissions/ref/System.Security.Permissions.csproj]: This is causing it to try to find a file System.Security.Permissions.netfx.cs, which does not exist. Maybe you need to conditionalize this to `'$(TargetsNetFx)' != 'true'` ? 

Good idea to `build -buildallconfigurations` to validate this kind of change locally (when you have changes that affect specific configurations)

**Comment by danmosemsft** [comment]
UAP is failing with
```
D:\j\workspace\windows-TGrou---33cbf18b\src\shims\shims.proj(45,5): warning : The type 'T:System.Security.IStackWalk' is defined in multiple seed assemblies. If this is intentional, specify one of the following arguments to choose the preferred seed type:  /preferSeedType:System.Security.IStackWalk=System.Runtime.Extensions  /preferSeedType:System.Security.IStackWalk=System.Security.Permissions
D:\j\workspace\windows-TGrou---33cbf18b\src\shims\shims.proj(45,5): warning : The type 'T:System.Security.Permissions.PermissionState' is defined in multiple seed assemblies. If this is intentional, specify one of the following arguments to choose the preferred seed type:  /preferSeedType:System.Security.Permissions.PermissionState=System.Runtime.Extensions  /preferSeedType:System.Security.Permissions.PermissionState=System.Security.Permissions
D:\j\workspace\windows-TGrou---33cbf18b\src\shims\shims.proj(45,5): warning : The type 'T:System.Security.PermissionSet' is defined in multiple seed assemblies. If this is intentional, specify one of the following arguments to choose the preferred seed type:  /preferSeedType:System.Security.PermissionSet=System.Runtime.Extensions  /preferSeedType:System.Security.PermissionSet=System.Security.Permissions
```

**Comment by maryamariyan** [comment]
I'll push my latest local changes in a few minutes

File: src/System.Security.Permissions/src/System.Security.Permissions.csproj [hunk 4]
```diff
@@ -172,6 +169,11 @@
     <Compile Include="System\Transactions\DistributedTransactionPermission.cs" />
     <Compile Include="System\Transactions\DistributedTransactionPermissionAttribute.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetGroup)' == 'netcoreapp' or '$(TargetGroup)' == 'uap'">
```
Comment (jkotas) [inline src/System.Security.Permissions/src/System.Security.Permissions.csproj]: These types should be included in the netstandard version only. For uap and netcoreapp, they are in System.Runtime.Extensions with this change.

Comment (maryamariyan) [reply src/System.Security.Permissions/src/System.Security.Permissions.csproj]: Thanks will update

File: src/System.Security.Permissions/ref/System.Security.Permissions.csproj [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
-    <IsPartialFacadeAssembly Condition="'$(TargetsNetFx)' == 'true'">true</IsPartialFacadeAssembly>
-    <Configurations>net461-Debug;net461-Release;netfx-Debug;netfx-Release;netstandard-Debug;netstandard-Release</Configurations>
+    <IsPartialFacadeAssembly Condition="'$(TargetsNetFx)' == 'true' or '$(TargetGroup)' == 'netcoreapp' or '$(TargetGroup)' == 'uap'">true</IsPartialFacadeAssembly>
+    <Configurations>net461-Debug;net461-Release;netfx-Debug;netfx-Release;netstandard-Debug;netstandard-Release;netcoreapp-Debug;netcoreapp-Release;uap-Debug;uap-Release</Configurations>
   </PropertyGroup>
   <ItemGroup>
     <Compile Include="System.Security.Permissions.cs" />
+    <Compile Include="System.Security.Permissions.netstandard.cs" Condition="'$(TargetsNetFx)' != 'true' and '$(TargetGroup)' == 'netstandard'" />
+    <Compile Include="System.Security.Permissions.netcoreapp.cs" Condition="'$(TargetGroup)' == 'netcoreapp'" />
+    <Compile Include="System.Security.Permissions.netcoreapp.cs" Condition="'$(TargetGroup)' == 'uap'" />
```
Comment (maryamariyan) [inline src/System.Security.Permissions/ref/System.Security.Permissions.csproj]: I'll clean up this part after I find the correct condition to fix the build issues

**Comment by maryamariyan** [comment]
@dotnet-bot test Linux x64 Release Build please

**Comment by maryamariyan** [comment]
@dotnet-bot test Linux x64 Release Build please

**Comment by maryamariyan** [comment]
@dotnet-bot test Linux x64 Release Build please

```
07:21:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] **Will try again after 5 min 0 sec**
[Pipeline] [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] {
[Pipeline] [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] httpRequest
07:26:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] HttpMethod: GET
07:26:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] URL: https://helix.dot.net/api/2017-04-14/jobs/4e82089f-0c8a-4a74-bde5-c258d054ae27/details
07:26:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Sending request to url: https://helix.dot.net/api/2017-04-14/jobs/4e82089f-0c8a-4a74-bde5-c258d054ae27/details
07:26:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Response Code: HTTP/1.1 200 OK
07:26:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Response: 
07:26:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] {"FailureReason":null,"QueueId":"Debian.8.Amd64.Open","JobList":"https://dotnetbuilddrops.blob.core.windows.net/build-8a92239944f2410fb6c575495cfcca9c/TestList.json?sv=2015-04-05&sr=c&sig=V%2Bp8LCAlmG97PEVixpKPt49eWjGMFOjQTy2ai1FtCuw%3D&st=2018-11-18T04%3A34%3A56.1906614Z&se=2018-12-18T04%3A34%3A56.1906639Z&sp=r","WorkItems":{"Unscheduled":0,"Waiting":208,"Running":0,"Finished":1,"ListUrl":"https://helix.dot.net/api/2017-04-14/jobs/4e82089f-0c8a-4a74-bde5-c258d054ae27/workitems"},"Name":"4e82089f-0c8a-4a74-bde5-c258d054ae27","Creator":"maryamariyan","Created":"2018-11-18T04:35:10.1320000+00:00","Finished":null,"InitialWorkItemCount":208,"WaitUrl":"https://helix.dot.net/api/2017-04-14/jobs/4e82089f-0c8a-4a74-bde5-c258d054ae27/wait","Source":"pr/jenkins/dotnet/corefx/master/","Type":"test/functional/cli/","Build":"07ae6a3991c34a4ec7b415ef3326a1d923bd25d4","Properties":{"architecture":"x64","configuration":"Release","operatingSystem":"Debian.8.Amd64.Open"},"Errors":[]}
07:26:45 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Success code from [100‥399]
[Pipeline] [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] httpRequest
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] HttpMethod: GET
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] URL: https://helix.dot.net/api/2017-04-14/aggregate/jobs?groupBy=job.name&maxResultSets=1&filter.name=4e82089f-0c8a-4a74-bde5-c258d054ae27
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Sending request to url: https://helix.dot.net/api/2017-04-14/aggregate/jobs?groupBy=job.name&maxResultSets=1&filter.name=4e82089f-0c8a-4a74-bde5-c258d054ae27
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Response Code: HTTP/1.1 200 OK
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Response: 
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] [{"Key":{"job.name":"4e82089f-0c8a-4a74-bde5-c258d054ae27"},"Data":{"Analysis":[{"Name":"xunit","Status":{"pass":0,"passonretry":0,"fail":0,"skip":0}}],"WorkItemStatus":{"run":208,"pass":1}}}]
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Success code from [100‥399]
[Pipeline] [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] echo
07:26:46 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] Info: Passed 0 (0 skipped)
[Pipeline] [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] }
07:26:47 [Debian.8.Amd64.Open - x64 - Release (4e82089f-0c8a-4a74-bde5-c258d054ae27)] **Will try again after 5 min 0 sec**
```

**Comment by MattGal** [comment]
@maryamariyan my apologies for the inconvenience, the problem should currently be mitigated, and we're actively working to address this now.  

Tracked by https://github.com/dotnet/core-eng/issues/4675

**Comment by maryamariyan** [comment]
@danmosemsft @safern @ericstj @jkotas this PR is ready for your review.

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.netcoreapp.cs [hunk 6]
```diff
@@ -21,7 +22,18 @@ public void GetSetupInformation()
...
+        }
+
+        [Fact]
+        public static void GetPermissionSet()
+        {
+            RemoteInvoke(() => {
+                PermissionSet permissionSet = AppDomain.CurrentDomain.PermissionSet;
+                Assert.NotNull(permissionSet);
+                Assert.Equal(permissionSet, AppDomain.CurrentDomain.PermissionSet);
+                return SuccessExitCode;
```
Comment (jkotas) [inline src/System.Runtime.Extensions/tests/System/AppDomainTests.netcoreapp.cs]: Assert that the returned permission set is unrestricted?

File: src/System.Security.Permissions/ref/System.Security.Permissions.netstandard.netfx.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+namespace System.Security
```
[the remark this answers is not in the document]
Comment (safern) [reply src/System.Security.Permissions/ref/System.Security.Permissions.netstandard.netfx.cs]: We've used it in some places, the most common obviously is ifdefs, but for example in SqlClient we use it: https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/ref/System.Data.SqlClient.csproj#L13

I don't have an strong opinion though, I usually use ifdefs for the ref.

**Comment by maryamariyan** [comment]
This PR is ready for review again.

**Comment by jkotas** [comment]
The C# delta looks fine to me. 

I do not know whether the packaging changes and the new dependencies that this is introducing are all fine. @ericstj Could you please take a look at those?

File: src/System.Security.Permissions/ref/System.Security.Permissions.cs [hunk 8]
```diff
@@ -573,13 +578,15 @@ public partial interface ISecurityPolicyEncodable
         void FromXml(System.Security.SecurityElement e, System.Security.Policy.PolicyLevel level);
         System.Security.SecurityElement ToXml(System.Security.Policy.PolicyLevel level);
     }
+#if netstandard || netfx
```
Comment (safern) [inline src/System.Security.Permissions/ref/System.Security.Permissions.cs]: You're missing net461 which is actually the assembly carried over in the package for full framework. So you would either need to do a `!netcoreapp && !uap` or add a DefineConstants entry for net461 here: https://github.com/dotnet/corefx/blob/master/Directory.Build.props#L418

 then just do `netstandard || netfx || net461`

File: src/System.Security.Permissions/src/System.Security.Permissions.csproj [hunk 9]
```diff
@@ -172,6 +169,11 @@
     <Compile Include="System\Transactions\DistributedTransactionPermission.cs" />
     <Compile Include="System\Transactions\DistributedTransactionPermissionAttribute.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetGroup)' == 'netstandard'">
+    <Compile Include="..\..\System.Runtime.Extensions\src\System\Security\IStackWalk.cs" />
```
Comment (safern) [inline src/System.Security.Permissions/src/System.Security.Permissions.csproj]: Should we instead move this files to common? That is the pattern we follow when shared across multiple projects.

Comment (jkotas) [reply src/System.Security.Permissions/src/System.Security.Permissions.csproj]: We have 250+ places where we do ` <Compile Include="..\..\`.

I think Common make sense for things that are really common within netcoreapp.

I do not think Common makes sense to share files for compat packages. These compat packages are transient. We are going to be building them for long - we dropped many of them already. Once we drop them, we would end up with stuff in Common that is not shared by anything.

**Comment by maryamariyan** [comment]
@dotnet-bot test OSX x64 Debug Build
@dotnet-bot test Linux x64 Release Build

**Comment by maryamariyan** [comment]
@ericstj do you approve the changes in this PR?

**Pull request merged**